File: parallel_plane_waveguide/python/IW_Parallel_Plate_Waveguide.py
# -*- coding: utf-8 -*-
"""
 Rectangular Waveguide Tutorial

 Tested with
  - python 3.10
  - openEMS v0.0.35+

 (c) 2015-2023 Thorsten Liebig <thorsten.liebig@gmx.de>

"""

### Import Libraries
import os, tempfile
from pylab import *

from CSXCAD.CSXCAD  import ContinuousStructure
from openEMS import openEMS

APPCSXCAD_CMD = '~/opt/openEMS/bin/AppCSXCAD'

### Setup the simulation
Sim_Path = os.path.join(os.getcwd(), 'sinusoidal')
print(f"SIM_PATH: {Sim_Path}")
###############################################################################################
###################################### SET CONSTANTS ###########################################
###############################################################################################
from openEMS.physical_constants import *

unit = 1 # drawing unit in mm, should be passed whenever creating a grid

### FREQUENCY
f0 = 10e6
fc = 10e3

lambda0 = (C0/f0)/unit; # Number of discrete steps for 1 wavelength


mesh_res = [1, 1, 1]
### PLATE DIMENSIONS
plate_x = np.linspace(-10, 10, 21)
plate_y = np.linspace(-10, 10, 21)
plate_z = np.linspace(-10, 30, 41)

###############################################################################################
###################################### INITIALIZE FDTD ########################################
###############################################################################################

### Setup FDTD parameter & excitation function
NrTS_ = 100
FDTD = openEMS(NrTS=NrTS_, EndCriteria=0, OverSampling=50)

#############################################################################################
################################# BOUNDARY CONDITIONS #######################################
#############################################################################################

# boundary conditions
FDTD.SetBoundaryCond(['PMC', 'PMC', 'PEC', 'PEC', 'MUR', 'MUR'])


#####################################################################################
################################# GRID / MESH #######################################
#####################################################################################

csx = ContinuousStructure()
FDTD.SetCSX(csx)

mesh = csx.GetGrid()
mesh.SetDeltaUnit(unit)


mesh.AddLine('x', plate_x)
mesh.AddLine('y', plate_y)
mesh.AddLine('z', plate_z)

#! WARNING: this is a guess
# mesh.SmoothMeshLines('all', (C0 / fc / unit) / 20, 1)
print(mesh)

####################################################################################
################################# EXCITATION #######################################
####################################################################################

FDTD.SetSinusExcite(f0)

'''
It seems like there is something wrong here with the voltage excitation
- Voltage excitations     : 420    (0, 420, 0)
vs
Warning: Unused primitive (type: Box) detected in property: excitation!
- Voltage excitations     : 0      (0, 0, 0)
'''
#! WARNING: IT SEEMS LIKE THE EXCITATION IS NOT ACTUALLY ACTIVE

# apply the excitation & resist as a current source
start = [plate_x[0], plate_y[0], 0]
stop  = [plate_x[-1], plate_y[-1], 0]

csx_excitation = csx.AddExcitation('excitation', 0, [0, 1, 0])
csx_excitation.AddBox(start, stop)

# #? edges2grid?
# feed_R = 0
# port = FDTD.AddLumpedPort(1, feed_R, start, stop, 'y', 1.0, priority=0)#, edges2grid='xy')

####################################################################################
#################################### DUMPS #########################################
####################################################################################

### Define dump box...
start = [plate_x[0], 0, plate_z[0]]
stop  = [plate_x[-1], 0, plate_z[-1]]

Et = csx.AddDump('Et')
Et.AddBox(start, stop)

### Run the simulation
CSX_file = os.path.join(Sim_Path, 'plane_waveguide.xml')
if not os.path.exists(Sim_Path):
    os.mkdir(Sim_Path)


csx.Write2XML(CSX_file)

from optparse import OptionParser
import inspect
results = inspect.getmembers(csx, predicate=inspect.isfunction)
print(results)
os.system(APPCSXCAD_CMD + ' "{}"'.format(CSX_file))
str_cmd = APPCSXCAD_CMD + ' "{}"'.format(CSX_file)
print(str_cmd)

#### CHECK DIMENSIONS
FDTD.Run(Sim_Path, cleanup=True, debug_material=True, debug_pec=True, debug_operator=True, debug_boxes=True, debug_csx=True, verbose=3)

#######################################################################################
#################################### COMMENTS #########################################
#######################################################################################


'''
### EXCITATION
For some reason the excitation is not passed on.
Volt_Count = volt_vIndex[0].size();
vector<unsigned int> volt_vIndex[3];

volt_vIndex[0].push_back(pos[0]);
virtual unsigned int GetNumberOfLines(int ny, bool full=false) const {UNUSED(full);return numLines[ny];}
-> REASON WAS NON-OVERLAPPING BOUNDARIES FOR EXCITATION SOURCE AND GRID
'''

File: tutorials/octave_tutorials/IW_Parallel_Plate_Waveguide.m
#############################################################################################
##################################### PATHS ADDED ###########################################
#############################################################################################

addpath('~/opt/openEMS/share/openEMS/matlab');
addpath('~/opt/openEMS/share/CSXCAD/matlab');
addpath('~/opt/openEMS/share/hyp2mat/matlab');

close all
clear
clc

save_folder = 'sinusoidal';


###############################################################################################
###################################### SET CONSTANTS ###########################################
###############################################################################################

physical_constants;
unit = 1e-3; % drawing unit in mm, should be passed whenever creating a grid

### FREQUENCY
f_start = 1e9;
f0 = 1.5e9;
f_stop = 2e9;
lambda0 = (c0/f0)/unit; % Number of discrete steps for 1 wavelength
display("wavelength per mm");
display(lambda0);
% lambda0 is 200 mm for 1.5 GHz -> So make the plates x: 2 times lambda and the Z: 4 times lambda
% and the y 1 times lambda in both directions
mesh_res = lambda0 ./ [30 30 30] % Make sure there are 30 mesh-steps per wavelength

% 6 mm per grid: 6*30 = 180 is 1 wavelength

### PLATE DIMENSIONS (in units)
plate_x = [-200 200];
plate_y = [-100 100];
plate_z = [-600 600];

###############################################################################################
###################################### INITIALIZE FDTD ###########################################
###############################################################################################

% init and define FDTD parameter
# 50: force given timestep?
% init and define FDTD parameter: oversampled at a factor 50 x the nyquist rate

%{
<<< DEFAULT VALUES >>>
- NrTS: number of time samples
%}
FDTD = InitFDTD('NrTS', 100, 'EndCriteria', 0, 'OverSampling',50);

#############################################################################################
################################# BOUNDARY CONDITIONS #######################################
#############################################################################################

# ---> [xmin, xmax, ymin, ymax, zmin, zmax] <---
# Parallel plate waveguide: on ymin and ymax there's a PEC.
# Excitation happens in the y-direction, in the xy plane.
% BC = {'PMC' 'PMC' 'PEC' 'PEC' 'MUR' 'MUR'};
BC = {'PMC' 'PMC' 'PEC' 'PEC' 'MUR' 'MUR'};
FDTD = SetBoundaryCond(FDTD,BC);

#####################################################################################
################################# GRID / MESH #######################################
#####################################################################################

% init and define FDTD mesh
CSX = InitCSX();
# Mesh goes from -10 -> 10
% SmoothMeshLines( lines, max_res, ratio, varargin)

% Distance between plates:
display("Mesh Resolution:");
display(mesh_res);

mesh.x = SmoothMeshLines(plate_x, mesh_res(1));
mesh.y = SmoothMeshLines(plate_y, mesh_res(2));
mesh.z = SmoothMeshLines(plate_z, mesh_res(3));

CSX = DefineRectGrid(CSX, unit, mesh);

####################################################################################
################################# EXCITATION #######################################
####################################################################################

%{
### Q: CHANGING EXCITATION TO HIGHER FREQUENCY
Make sure your mesh is small enough
- 
%}

# 10 MHz Sinusoidal excitation
FDTD = SetSinusExcite(FDTD,f0); # SINUSOIDAL EXCITATION
% FDTD = SetGaussExcite(FDTD, 1e9, 1e6); # WAVELET EXCITATION

% define the excitation
% Excitation rectangle in the xy-plane.
start = [mesh.x(1), mesh.y(1), 0];
stop = [mesh.x(end), mesh.y(end), 0];

display("excitation: ")
display(start);
display(stop);
% 0: soft field excitation, [0 1 0]: excitation in y-direction?
CSX = AddExcitation(CSX, 'excitation', 0, [0 1 0]);
% 0: soft field excitation, start, stop -> where the excitation starts and where it stops
% Question: how do I know the material of the grid?
CSX = AddBox(CSX,'excitation', 0, start, stop);

####################################################################################
#################################### DUMPS #########################################
####################################################################################

% define a time domain e-field dump box
start = [mesh.x(1) 0 mesh.z(1)];
stop = [mesh.x(end) 0 mesh.z(end)];

display("dump: ")
display(start);
display(stop);

CSX = AddDump(CSX,'Et','DumpMode',0);
CSX = AddBox(CSX,'Et',0, start, stop);

% CSX = AddDump(CSX,'Ht','DumpMode',1)
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

% CSX = AddDump(CSX,'It', 'DumpMode',2);
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

% CSX = AddDump(CSX, 'rotH', 'DumpMode',3);
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

% CSX = AddDump(CSX, 'Dt', 'DumpMode',4);
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

% CSX = AddDump(CSX, 'Bt', 'DumpMode',5);
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

% CSX = AddDump(CSX,'Ef','DumpMode',10);
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

% CSX = AddDump(CSX,'Hf','DumpMode',11);
% CSX = AddBox(CSX,'Et',0,[-10 0 -10],[10 0 30]);

# Add the electric field in the xz plane
# *Q*: Why would you need to add the elctric field here?
# - 

% remove old simulation results (if exist)
rmdir(save_folder,'s');mkdir(save_folder);

% write openEMS xml data file
WriteOpenEMS(strcat(save_folder,'/tmp.xml'),FDTD,CSX);

% view defined structure
CSXGeomPlot(strcat(save_folder,'/tmp.xml'));

% run openEMS simulation
RunOpenEMS(save_folder,'tmp.xml','');

disp('use Paraview to visualize the FDTD result...');

#######################################################################################
#################################### COMMENTS #########################################
#######################################################################################
%{
### LAUNCH
At launch don't forget to: "unset GTK_PATH", in order to be able to run octave

### PASSING PARAMETERS
The way openEMS works is the parameters coming always needs to be indicated by the "string-name" of the parameter.
e.g.: "oversampling", 50 -> means oversampling needs to be done at a multiple of 50 of the nyquist frequency.


### SIMULATION TOO SLOW
Try to reduce the simulation grid, excitation etc.. size.

### GRID SIZE INCORRECT
The grid size for some reason shows only 12 or 6 lines. The gridding isn't correct.

### TIME STEP INCORRECT
The speed of light is: 299792458 m/s, 
- Which means 1 m is travelled in 3.33564095e-9, and half of that for 500 mm
So obviously the length of the timestep needs to be smaller, or the grid needs to be larger than that.
The timestep seemed to be needs to be obviously smaller than 
FDTD timestep is: 1.92308e-09 s; Nyquist rate: 25 timesteps @1.04e+07 Hz

### ERROR RETURN 137
Means that the linux memory maanger sent SIGKILL to your process because it used to much memory
== check "cat /var/log/syslog", which shows in this case
> Oct 15 16:22:47 Igor kernel: [21770.664249] Out of memory: Killed process 55834 (openEMS) total-vm:10425436kB, anon-rss:9135488kB, file-rss:224kB, shmem-rss:0kB, UID:1000 pgtables:18200kB oom_score_adj:0
Which shows 10.425436 Gb (more than 16 Gb we have)
%}